include/oneapi/dpl/internal/scan_by_segment_impl.h

```diff
@@ -0,0 +1,506 @@
+
+                auto __partials_acc = __partials.template get_access<sycl::access_mode::read>(__cgh);
+                auto __seg_ends_acc = __seg_ends.template get_access<sycl::access_mode::read>(__cgh);
+
+                __cgh.depends_on(__wg_incl_scan);
+
+                __cgh.parallel_for(
+                    sycl::nd_range<1>{__n_groups * __wgroup_size, __wgroup_size}, [=](sycl::nd_item<1> __item) {
+                        auto __group = __item.get_group();
+                        auto __group_id = __group.get_group_id(0);
+                        auto __global_id = __item.get_global_id();
+                        auto __local_id = __item.get_local_id();
+
+                        int __wg_agg_idx = __group_id - 1;
+                        __val_type __agg_collector{};
+                        bool __first = true;
+
+                        int __start = __global_id * __vals_per_item;
+                        int __end = sycl::minimum{}(__start + __vals_per_item, __n);
+
+                        // 2a.Check to see if an aggregate exists and compute 
+                        // that value in the first work item.
+                        bool __ag_exists = false;
+                        if (__local_id == 0 && __wg_agg_idx >= 0)
+                        {
+                            if (__start < __n && __keys[__start] == __keys[__start - 1])
+                            {
+                                __ag_exists = true;
+                                for (int32_t __i = __wg_agg_idx; __i >= 0; --__i)
+                                {
+                                    const auto& __wg_aggregate = __partials_acc[__i];
+                                    const auto& __b_seg_end = __seg_ends_acc[__i];
+
+                                    if (__first)
+                                    {
+                                        __agg_collector = __wg_aggregate;
+                                        __first = false;
+                                    }
+                                    else
+                                        __agg_collector = __binary_op(__wg_aggregate, __agg_collector);
+
+                                    // current aggregate is the last aggregate
+                                    if (__b_seg_end)
+                                        break;
+                                }
+                            }
+                        }
+
+                        __ag_exists = sycl::group_broadcast(__group, __ag_exists);
+                        if (!__ag_exists)
+                            return;
+
+                        __agg_collector = sycl::group_broadcast(__group, __agg_collector);
+
+                        // 2c. Second pass over the keys, reidentifying end segments and applying work group
+                        // aggregates if appropriate.
+                        ::std::size_t __local_min_key_idx = __n - 1;
```

Comment:
May we can execute this line for case when __n == 0 then we will have an error, I think.

Reply:
Yes, this would be an error. I will change this is to a signed type.